infra/src/Pos/Infra/Network/DnsDomains.hs

```diff
@@ -38,7 +62,111 @@ data NodeAddr a =
+
+instance ToJSON (NodeAddr (Maybe DNS.Domain)) where
+    toJSON =
+        A.object . \case
+            (NodeAddrExact ip mWord) ->
+                case (ip, mWord) of
+                    (IPv4 (IP4 ip4), Just port) ->
+                        [ "addr" .= (A.String . show $ IPv4 (IP4 ip4))
+                        , "port" .= (toJSON port)
+                        ]
+                    (IPv4 (IP4 ip4), Nothing)   ->
+                        [ "addr" .= (A.String . show $ IPv4 (IP4 ip4))
+                        , "port" .= (toJSON (Prelude.read "3000" :: Word16))
+                        ]
+                    (IPv6 (IP6 ip6), Just port) ->
+                        [ "addr" .= (A.String . show $ IPv6 (IP6 ip6))
+                        , "port" .= (toJSON port)
+                        ]
+                    (IPv6 (IP6 ip6), Nothing)   ->
+                        [ "addr" .= (A.String . show $ IPv6 (IP6 ip6))
+                        , "port" .= (toJSON (Prelude.read "3000" :: Word16))
+                        ]
+            (NodeAddrDNS nHost mWord) ->
+                case (nHost, mWord) of
+                    (Just host, Just port) ->
+                        [ "host" .= (toJSON $ decodeUtf8 @Text host)
+                        , "port" .= (toJSON port)
+                        ]
+                    (Just host, Nothing)   ->
+                        [ "host" .= (toJSON $ decodeUtf8 @Text host)
+                        , "port" .= (toJSON (Prelude.read "3000" :: Word16))
+                        ]
+                    (Nothing, _)      -> error "Please enter hostname"
+
+instance FromJSON (NodeAddr (Maybe DNS.Domain)) where
+    parseJSON (A.Object o)
+        | (HMS.member "addr" o) && (HMS.member "port" o) = do
+              addrV <- o .: "addr"
+              addr <- parseJSON addrV
+              portV <- o .: "port"
+              port <- parseJSON portV
+              return $ NodeAddrExact addr port
+        | (HMS.member "host" o) && (HMS.member "port" o) = do
+              portV <- o .: "port"
+              port <- parseJSON portV
+              hostV <- o .: "host"
+              case hostV of
+                  A.String text -> do
+                      let host = encodeUtf8 @Text @ByteString text
+                      return $ NodeAddrDNS (Just host) port
+                  _             -> fail "Incorrect JSON encoding for hostname"
+        | otherwise =
+            error "Incorrect JSON encoding for NodeAddr (Maybe DNS.Domain)"
+    parseJSON invalid = A.typeMismatch "NodeAddr (Maybe DNS.Domain)" invalid
+
+-- | Orphan instance; lifted from aeson-iproute
+instance ToJSON IP where
+    toJSON = A.String . T.pack . show
+
+-- | Orphan instance; lifted from aeson-iproute
+instance FromJSON IP where
+    parseJSON (A.String s)
+        | Just r <- readMaybe (T.unpack s) = pure r
+        | otherwise = fail "Unable to parse IP"
+    parseJSON v = A.typeMismatch "IP" v
+
+instance ToJSON (NodeAddr DNS.Domain) where
```

Comment:
@intricate yeah I definitely agree, I must've skimmed over this and missed it.

The instances for `ToJSON (NodeAddr DNS.Domain)` and `ToJSON (NodeAddr (Maybe DNS.Domain))` seem to be basically identical, apart from the `Nothing` case. It would be great to refactor this to use a common function for the core.